repo_remote.go

```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
 	return RepoRemoveRemote(r.path, name, opts...)
 }
+
+// RemotesListOptions contains arguments for listing remotes of the repository.
+// Docs: https://git-scm.com/docs/git-remote#_commands
+type RemotesListOptions struct {
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemotesList lists remotes of the repository in given path.
+func RepoRemotesList(repoPath string, opts ...RemotesListOptions) ([]string, error) {
+	var opt RemotesListOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	stdout, err := NewCommand("remote").RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return stdoutToStringSlice(stdout), nil
+}
+
+// RemotesList lists remotes of the repository.
+func (r *Repository) RemotesList(opts ...RemotesListOptions) ([]string, error) {
+	return RepoRemotesList(r.path, opts...)
+}
+
+// RemoteURLGetOptions contains arguments for retrieving URL(s) of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLGetOptions struct {
+	// False: get fetch URLs
+	// True: get push URLs
+	Push bool
+	// True: get all URLs (lists also non-main URLs; not related with Push)
+	All bool
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemoteURLGet retrieves URL(s) of a remote of the repository in given path.
+func RepoRemoteURLGet(repoPath, name string, opts ...RemoteURLGetOptions) ([]string, error) {
+	var opt RemoteURLGetOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "get-url")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+	if opt.All {
+		cmd.AddArgs("--all")
+	}
+
+	stdout, err := cmd.AddArgs(name).RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil {
+		return nil, err
+	}
+	return stdoutToStringSlice(stdout), nil
+}
+
+// RemoteURLGet retrieves URL(s) of a remote of the repository in given path.
+func (r *Repository) RemoteURLGet(name string, opts ...RemoteURLGetOptions) ([]string, error) {
+	return RepoRemoteURLGet(r.path, name, opts...)
+}
+
+// RemoteURLSetOptions contains arguments for setting an URL of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLSetOptions struct {
+	// False: set fetch URLs
+	// True: set push URLs
+	Push bool
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemoteURLSetFirst sets first URL of the remote with given name of the repository in given path.
+func RepoRemoteURLSetFirst(repoPath, name, newurl string, opts ...RemoteURLSetOptions) error {
```

Comment:
How is the "first" come into play? Not seeing "first" is used as git args?

Reply:
uhh yeah so.. repo can have like
origin urls:
 - type fetch url: foobar1
 - type push url: foobar2
 - type fetch url: foobar3
 - type fetch url: foobar4
 
^^^ in this case, the first fetch url is foobar1, and first push url is foobar3

In reality, this almost never happens, and even if it does, the 'first' is used elsewhere as well.

It might be a good idea to add a short 'it is almost always first, and if not, you probably don't have to worry', but don't know on how to word it.

Reply:
Thanks for the context!

Either way I think we do not need explicitly put "first" in our function name because we are basically pass-on the value from Git CLI, and whatever it returns is what we return to the caller.